### PR TITLE
EWS - Adding O365 Tenant Allow/Block List Management

### PR DESCRIPTION
**Pull request by mikejrizzo**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
No related issues.

## Description
This pull request adds additional commands to the "EWS Extension Online Powershell v2" to allow management of O365 Tenant Allow/Block Lists.  This O365 feature allows for simple tenant-wide blocking or allowing of things such as email senders, file hashes, and URLs.  This feature is managed using the same PowerShell module already incorporated into the integration.

Existing functionality of this integration is not changed.  This contribution only adds new functionality.

Official documentation of this O365 feature can be found here: https://docs.microsoft.com/en-us/microsoft-365/security/office-365-security/tenant-allow-block-list?view=o365-worldwide

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- 5.5 or later - same as the current published version of this integration.

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ X ] No

## Must have
- [ ] Tests (Published integration contains no tests)
- [ X ] Documentation

### CODE REVIEW

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml [hunk 1]
```diff
@@ -363,15 +278,310 @@ script:
...
-      and distribution groups).
-    execution: false
+    description: Displays existing recipient objects in your organization. This command returns all mail-enabled objects (for example, mailboxes, mail users, mail contacts, and distribution groups).
     name: ews-recipient-list
-  isfetch: false
+  - name: ews-new-tenant-allow-block-list-items
+    arguments:
+    - name: entries
+      required: true
+      description: Entries to add to the list.  Separate multiple entries with a comma (e.g. "Item1,Item2").
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:287]:
```suggestion
      description: A comma-separated list of entries to add to the list (e.g. "Item1,Item2").
```

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml [hunk 2]
```diff
@@ -363,15 +278,310 @@ script:
...
+      - sender
+      - filehash
+      - url
+      description: List type to add items to.
+    - name: list_subtype
+      auto: PREDEFINED
+      predefined:
+      - Tenant
+      - AdvancedDelivery
+      description: 'List subtype to add items to. '
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:302]:
```suggestion
      description: List subtype to add items to.
```

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml [hunk 3]
```diff
@@ -363,15 +278,310 @@ script:
...
+      required: true
+      auto: PREDEFINED
+      predefined:
+      - Block
+      - Allow
+      description: Action to set for new entries
+    - name: notes
+      description: Notes to include on new list entries
+    - name: expiration_date
+      description: Enter a specific date and time for the new entries to expire using format "YYYY-MM-DD HH:MM:SSz" for UTC time.  Alternately, a PowerShell **GetDate** statement can be used.
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:314]: Can we use the ISO format (`YYYY-MM-DDTHH:MM:SSZ`) instead? We are using it across the product.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml [hunk 4]
```diff
@@ -363,15 +278,310 @@ script:
...
+      defaultValue: Tenant
+    - name: action
+      required: true
+      auto: PREDEFINED
+      predefined:
+      - Block
+      - Allow
+      description: Action to filter entries by.
+    - name: expiration_date
+      description: Enter a specific date and time to filter entries by using format "YYYY-MM-DD HH:MM:SSz" for UTC time.  Alternately, a PowerShell **GetDate** statement can be used.
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:404]: same here

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1 [hunk 5]
```diff
@@ -398,6 +576,125 @@ function GetEXOCASMailboxCommand
...
+    [CmdletBinding()]
+    Param (
+        [Parameter(Mandatory)][ExchangeOnlinePowershellV2Client]$client,
+        [hashtable]$kwargs
+    )
+    $entries = $kwargs.entries
+    $ids = $kwargs.ids
+    $list_type = $kwargs.list_type
+    $list_subtype = $kwargs.list_subtype
+    if ($entries -and $ids)
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1:678]: What if not are passed?

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1 [hunk 6]
```diff
@@ -398,6 +576,125 @@ function GetEXOCASMailboxCommand
...
+    }
+    $raw_response = $client.EXORemoveTenantAllowBlockList(
+        $entries, $ids, $list_type, $list_subtype
+    )
+    $human_readable = TableToMarkdown $raw_response "Results of $command"
+    if ($kwargs.set_context -eq "yes")
+    {
+        $entry_context = @{ "$script:INTEGRATION_ENTRY_CONTEXT.RemovedTenantBlocks" = $raw_response }
+        Write-Output $human_readable, $entry_context, $raw_response
+    }
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1:690]: No need to write to context on "delete" commands

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1 [hunk 7]
```diff
@@ -398,6 +576,125 @@ function GetEXOCASMailboxCommand
...
+    $list_subtype = $kwargs.list_subtype
+    $action = $kwargs.action
+    $notes = $kwargs.notes
+    $no_expiration = if ($kwargs.no_expiration -eq "true") { $true } else { $false }
+    $expiration_date = $kwargs.expiration_date
+    $raw_response = $client.EXONewTenantAllowBlockList(
+        $entries, $list_type, $list_subtype, $action, $notes, $no_expiration, $expiration_date
+    )
+    $human_readable = TableToMarkdown $raw_response "Results of $command"
+    if ($kwargs.set_context -eq "yes")
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1:598]: The command should always set the context.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1 [hunk 8]
```diff
@@ -398,6 +576,125 @@ function GetEXOCASMailboxCommand
...
+    $list_type = $kwargs.list_type
+    $list_subtype = $kwargs.list_subtype
+    $action = $kwargs.action
+    $no_expiration = if ($kwargs.no_expiration -eq "true") { $true } else { $false }
+    $expiration_date = $kwargs.expiration_date
+    $raw_response = $client.EXOGetTenantAllowBlockList(
+        $entry, $list_type, $list_subtype, $action, $no_expiration, $expiration_date
+    )
+    $human_readable = TableToMarkdown $raw_response "Results of $command"
+    if ($kwargs.set_context -eq "yes")
```
Comment (yaakovi) [inline Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1:627]: The command should always set the context.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2_description.md [hunk 9]
```diff
@@ -6,3 +6,7 @@ To use this integration, you need to connect an application with a certificate.
 2. Run the **CreateCertificate** command in Cortex XSOAR to acquire the certificate.
 3. Attach the .cer file to your Azure App.
 4. Copy the contents of the .txt file and paste it in the Certificate parameter of the integration's instance.
+
+
+---
+[View Integration Documentation](https://xsoar.pan.dev/docs/reference/integrations/ews-extension-online-powershell-v2)
```
[the remark this answers is not in the document]
Comment (mikejrizzo) [reply Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2_description.md:12]: I think this was added by one of the demisto-sdk commands I ran to format and validate files, but I've manually removed it.

Done.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml (hunk 1 above)
Comment (mikejrizzo) [reply Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:287]: Done.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml (hunk 2 above)
Comment (mikejrizzo) [reply Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:302]: Done.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml (hunk 3 above)
Comment (mikejrizzo) [reply Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:314]: The time format I used came from the Microsoft documentation for this Powershell cmdlet, however I tested the command with ISO formatting and that works as well.  I'm updating the field description to specify to use ISO format.

Done.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml (hunk 4 above)
Comment (mikejrizzo) [reply Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.yml:404]: Done.

File: Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1 (hunk 6 above)
Comment (mikejrizzo) [reply Packs/EWS/Integrations/EwsExtensionEXOPowershellV2/EwsExtensionEXOPowershellV2.ps1:690]: This is intended to allow storing a list of deleted items for later confirmation in playbooks/scripts that deletion was successful.  I would suggest keeping this here unless it cannot be included.

**Comment by mikejrizzo** [comment]
@yaakovi I've pushed the various changes we discussed in our meeting on Tuesday.  Please let me know if you have any further questions or items I need to look at.

Thanks!

**Review by yaakovi — APPROVED** [review]

**Pull request merged**